OSD-19689: implementing a place holder feature for notifications

Nikokolas3270 opened the pull request:
Place holders are replaced with the alert labels & annotations in the service log to generate

### What type of PR is this?
_feature_

### What this PR does / why we need it?
This PR allows to use alert labels & annotations in the service logs sent by `ocm-agent`

### Which Jira/Github issue(s) this PR fixes?

_Implements_ [OSD-19689](https://issues.redhat.com//browse/OSD-19689)

### Special notes for your reviewer:

### Pre-checks (if applicable):
- [X] Tested latest changes against a cluster
- [ ] Ran `make generate` command locally to validate code changes
- [X] Included documentation changes with PR

typeid commented on pkg/ocm/servicelog.go:
```diff
@@ -0,0 +1,88 @@
+)
+
+func (sl *ServiceLog) ReplacePlaceHoldersWithAlertLabelsAndAnnotations(alert *template.Alert) error {
+	replacePlaceHoldersInString := func(s string) (string, error) {
+		var err error
+		resolvePlaceHolder := func(placeHolder string) string {
+			if err == nil {
+				key, value, isOk := placeHolder[2:len(placeHolder)-1], "", false
+
+				if value, isOk = alert.Labels[key]; !isOk {
+					if value, isOk = alert.Annotations[key]; !isOk {
+						err = fmt.Errorf("alert has no '%s' label or annotation which could be used to replace place holders in the template", key)
+
+						return placeHolder
+					}
+				}
+				return value
+			}
+
+			return placeHolder
+		}
+
+		return slVarRefRe.ReplaceAllStringFunc(s, resolvePlaceHolder), err
+	}
```
**Question:** Is there a good reason to declare this function within `ReplacePlaceHoldersWithAlertLabelsAndAnnotations` and `resolvePlaceHolder` within `replacePlaceHoldersInString`? IMO this is pretty hard to read

Nikokolas3270 replied:
`resolvePlaceHolder` function changes the `err` variable in `replacePlaceHoldersInString` function.
For that reason `resolvePlaceHolder` function is nested in in `replacePlaceHoldersInString` function.

I would have preferred to avoid this side effect by the signature of `resolvePlaceHolder` has to be `func(string) string` when calling `slVarRefRe.ReplaceAllStringFunc`.

I initially designed the code by using [`FindStringSubmatchIndex`](https://pkg.go.dev/regexp#Regexp.FindStringSubmatchIndex) in a loop but it was not elegantly possible to deal with recursive variable replacement with that method.

I also though of using [`FindAllStringSubmatchIndex`](https://pkg.go.dev/regexp#Regexp.FindAllStringSubmatchIndex), but the code was then needlessly complicated.

typeid replied:
> resolvePlaceHolder function changes the err variable in replacePlaceHoldersInString function.
For that reason resolvePlaceHolder function is nested in in replacePlaceHoldersInString function.

This part makes sense. But can't we move `replacePlaceHoldersInString` out of `ReplacePlaceHoldersWithAlertLabelsAndAnnotations`?

Nikokolas3270 replied:
Technically we could, but `replacePlaceHoldersInString` uses the `alert` parameter passed to `ReplacePlaceHoldersWithAlertLabelsAndAnnotations`; if I move it out of it:
-  I will have to add some `alert` parameter to `replacePlaceHoldersInString`
- By doing so I may have to rename the method to tell that the place holders are replaced with the labels & annotations taken from the new `alert` parameter... while this was implicit with the previous nesting. Method name could become `replacePlaceHoldersWithAlertLabelsAndAnnotationsInString`... this is long, not sure this is suitable.

typeid commented on pkg/ocm/servicelog.go:
```diff
@@ -0,0 +1,88 @@
+		return slVarRefRe.ReplaceAllStringFunc(s, resolvePlaceHolder), err
+	}
+
+	var err error
+
+	if sl.Summary, err = replacePlaceHoldersInString(sl.Summary); err != nil {
+		return err
+	}
```
Would save two lines? :) 

```suggestion
	if sl.Summary, err := replacePlaceHoldersInString(sl.Summary); err != nil {
		return err
	}
```

Nikokolas3270 replied:
I think that `:=` declare and define/set a variable...
... and `sl.Summary` cannot be declared as it is already declared by being a member of `sl`.

`=` does not have this problem.

typeid replied:
`:=` works here as well, as it works if only one of the variables needs a fresh initialization. It would only initialize `err` and replace `sl.Summary`

Nikokolas3270 replied:
`:=` only works for variables, not for members:
https://go.dev/play/p/5yJGO_eVEZ8

typeid replied:
TIL. Thanks! :)

typeid commented on pkg/ocm/servicelog.go:
```diff
@@ -0,0 +1,88 @@
+	replacePlaceHoldersInString := func(s string) (string, error) {
+		var err error
+		resolvePlaceHolder := func(placeHolder string) string {
+			if err == nil {
```
What's the  `err == nil` meaning here? When is it not nil, if I'm not mistaken, it seems to just take the nil declaration from above?

typeid commented on pkg/metrics/metrics.go:
```diff
@@ -138,6 +145,14 @@ func CountServiceLogSent(template, state string) {
 }
 
+// CountFailedServiceLogs counts the total number of failed service logs (by notification template)
+func CountFailedServiceLogs(template string) {
```
Good work on adding this additional metric :)

bmeng commented on pkg/metrics/metrics.go:
```diff
@@ -17,7 +17,7 @@ var (
 	metricRequestsByService = prometheus.NewCounterVec(
 		prometheus.CounterOpts{
-			Name: "ocm_agent_requests_by_service",
+			Name: "ocm_agent_requests_by_service", // this is a counter without the _total suffix. What about renaming metric to ocm_agent_requests_by_service_total ?
```
Would you like to apply all the changes on the metrics naming suggestion? We can review after that.
But I think if we change the metrics name, we should update all the existing references, like prometheusrules if there is any.
And I remember we have added those metrics to the grafana dashboard, not sure if any update there needed.

Nikokolas3270 replied:
> Would you like to apply all the changes on the metrics naming suggestion?

Maybe, but not in this pull request as the only purpose of this PR is to add the place holder feature to `ocm-agent`

> But I think if we change the metrics name, we should update all the existing references, like prometheusrules if there is any.
And I remember we have added those metrics to the grafana dashboard, not sure if any update there needed.

Indeed, that's quite some work... also the `PrometheusRule`s and dashboards will probably have to be modified to accommodate both old & new names for a while.
So, to be honest, I don't know if it would be that much of a great idea to rename those metrics.
However I think this should be done in a separate pull request to not clutter and delay this PR which is quite urgent.
But before even submitting a PR for this change, I think it would be best to create a ticket and have the various contributors (team leads, team members, func lead) debate over the necessity of this change. @bmeng : WDYT?
My comments about the names of those metrics was more to open the debate :) ... and I can remove them if we create a ticket to track this renaming question.

bmeng replied:
+1 to put them into separated PR if required
Feel free to create a card to discuss with the owners firstly I would suggest

Nikokolas3270 replied:
@bmeng: I created a separate ticket for that: [OSD-19707](https://issues.redhat.com/browse/OSD-19707) :-)

Nikokolas3270 commented on pkg/handlers/helper.go:
```diff
@@ -59,7 +55,7 @@ type AMReceiverAlert template.Alert
 //go:generate mockgen -destination=mocks/helper.go -package=mocks github.com/openshift/ocm-agent/pkg/handlers OCMClient
 type OCMClient interface {
-	SendServiceLog(summary, firingDesc, resolveDesc, clusterID string, severity v1alpha1.NotificationSeverity, logType string, references []v1alpha1.NotificationReferenceType, firing bool) error
+	SendServiceLog(sl *ocm.ServiceLog) error
```
Hello, I am not sure of the urgency of this change (i.e. wether or not we could wait on PR https://github.com/openshift/ocm-agent/pull/66), raising the question to the epic owner.

Nikokolas3270 reviewed (commented):
@bmeng : I removed the comments on the metrics names.

bmeng commented:
@Nikokolas3270 The #66 merged, please help rebase here

Nikokolas3270 commented:
@bmeng: as discussed, this PR has been adapted to take into account the merge of [PR#66](https://github.com/openshift/ocm-agent/pull/66) which was about changing the way we sent SL to OCM.

I also took the liberty of fixing the 2 following problems I discovered with PR#66:
- Service log description was suffixed with `References: null` is case of `nil` references; that's no longer the case.
- `OCMClient` interface mock was no longer generated. I added back the comment considered by `go generate` in order to be able to generate the mock gain when running `make mockgen`

Finally remark that I also removed the `ServiceLogsHandler` type. Indeed this type was too close to `OCMClient`; the only additional thing this type was doing on top of `OCMClient` was to format the service log; this is now outsourced in a `ServiceLogBuilder` class which is (indirectly) unit tested (c.c. @xiaoyu74 the author of PR#66).

xiaoyu74 commented:
> @bmeng: as discussed, this PR has been adapted to take into account the merge of [PR#66](https://github.com/openshift/ocm-agent/pull/66) which was about changing the way we sent SL to OCM.
> 
> I also took the liberty of fixing the 2 following problems I discovered with PR#66:
> 
> * Service log description was suffixed with `References: null` is case of `nil` references; that's no longer the case.
> * `OCMClient` interface mock was no longer generated. I added back the comment considered by `go generate` in order to be able to generate the mock gain when running `make mockgen`
> 
> Finally remark that I also removed the `ServiceLogsHandler` type. Indeed this type was too close to `OCMClient`; the only additional thing this type was doing on top of `OCMClient` was to format the service log; this is now outsourced in a `ServiceLogBuilder` class which is (indirectly) unit tested (c.c. @xiaoyu74 the author of PR#66).

@Nikokolas3270 - Great work for refactoring my previous PR#66 to separate the log construction from the sending logic with `ServiceLogBuilder`.  It's more clear and modularize.  Thanks.

Just one question why we introduced the function `NewTestServiceLog` considering you've created `ServiceLogBuilder` struct to build service log entries ?

Nikokolas3270 commented:
> Just one question why we introduced the function `NewTestServiceLog` considering you've created `ServiceLogBuilder` struct to build service log entries ?

Hello @xiaoyu74, thanks for your feedback. I considered (for a while) re-using the `ServiceLogBuilder` to create the expected `ServiceLog`, but by doing so:
- I wouldn't have fully tested the `ServiceLogBuilder` as the got & expected `ServiceLog`s would have gone through some shared code to be created.
- Also a new `Build` method (i.e. `BuildTest`) would have needed to be added to `ServiceLogBuilder`. Indeed when creating the expected service log, I know the exact expected summary & description. There should be no choosing between firing & resolving there.
- Similarly a `NewTestServiceLogBuilder` constructor would have been needed. Indeed for the expected service log, there is no need to have an `firingDesc` and a `resolveDesc`.
- All in all this would have cluttered `ServiceLogBuilder` and more code would have been needed.

xiaoyu74 commented:
/lgtm

ravitri commented on pkg/handlers/service_logs.go:
```diff
@@ -4,88 +4,157 @@ import (
+
+			if value, isOk = alert.Labels[key]; !isOk {
+				if value, isOk = alert.Annotations[key]; !isOk {
+					err = fmt.Errorf("alert has no '%s' label or annotation which could be used to replace place holders in the template", key)
+
+					return placeHolder
```
Can a test be added for this block too? Also, if am reading it right, did you mean `return err` on line 60?

ravitri replied:
Ah ok.. Got it.. :+1:

bmeng commented on pkg/handlers/service_logs.go:
```diff
@@ -4,88 +4,157 @@ import (
+	slVarRefRe = regexp.MustCompile(`\${[^{}]*}`)
+)
+
+func replacePlaceHoldersInString(s string, alert *template.Alert) (string, error) {
```
Is it only me who feels the function is hard to understand?...

Nikokolas3270 replied:
This function is indeed a bit hard to understand.
I could add some comment telling that the place holders in `s` are replaced with the given `alert` labels & annotations.

The body of the function is also a bit hard to read because of the inner function but the code is actually much shorter that way. Explicitly looping on the matched token is actually clumsier; especially if we want to avoid recursive replacements (and we want that).

bmeng commented:
/lgtm

bmeng commented:
/approve
